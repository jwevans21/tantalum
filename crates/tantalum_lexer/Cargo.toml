[package]
name = "tantalum_lexer"
version = "0.1.0"
authors.workspace = true
edition.workspace = true
license.workspace = true
readme.workspace = true
repository.workspace = true
publish.workspace = true

[dependencies]
serde = { workspace = true, optional = true }

tantalum_span = { path = "../tantalum_span" }

[features]
default = ["serde"]
serde = ["dep:serde"]

[dev-dependencies]
insta = { workspace = true}
pretty_assertions = { workspace = true }
rstest = { workspace = true }


[lints]
workspace = true
