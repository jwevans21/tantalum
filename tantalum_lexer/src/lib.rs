//! # Tantalum Lexer
//!
//! Consumes source code written in the tantalum language and produces tokens
//! to be consumed by a parser.

#[cfg(test)]
mod tests;

pub mod token;
pub mod token_kind;

use tantalum_span::{Location, Span, Spanned};

use crate::token::Token;
use crate::token_kind::TokenKind;

/// The lexer for the Tantalum language, this will consume source code and produce
/// tokens with origin information. This lexer is designed to be used in a streaming
/// fashion, where tokens are consumed as they are needed.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Lexer<'file_name, 'source> {
    /// The current location of the lexer
    location: Location<'file_name>,
    /// The source code to lexed into tokens
    source: &'source str,
}

impl<'file_name, 'source> Lexer<'file_name, 'source> {
    /// Create a lexer for the Tantalum language over the provided `source`
    ///
    /// The file name is provided to allow useful diagnostics to be produced
    ///
    /// # Example
    /// ```
    /// use tantalum_lexer::Lexer;
    ///
    /// let lexer = Lexer::new("main.tan", "fn main(): i32 { return 0; }");
    ///
    /// // Use the lexer here (will likely require `lexer` to be mutable)
    /// ```
    #[must_use]
    #[inline]
    pub fn new(file_name: &'file_name str, source: &'source str) -> Self {
        return Self {
            location: Location::new(file_name),
            source,
        };
    }

    #[must_use]
    #[inline]
    pub fn location(&self) -> Location<'file_name> {
        return self.location;
    }

    #[must_use]
    #[inline]
    pub fn source(&self) -> &'source str {
        return self.source;
    }

    #[must_use]
    #[inline]
    pub fn file_name(&self) -> &'file_name str {
        return self.location.file_name();
    }

    #[must_use]
    #[inline]
    #[expect(clippy::too_many_lines, reason = "Long match statement")]
    pub fn next_token(&mut self) -> Option<Spanned<'file_name, Token<'source>>> {
        self.skip_whitespace();

        let start = self.location;

        /// Produce a Token based on a type and length
        macro_rules! lex {
            // Special case  for tokens already consumed
            ($production:path, 0) => {{
                return self.create_token($production, start);
            }};
            // General case for tokens that need to be consumed
            ($production:path, $length:expr) => {{
                let _: char = self.next_characters($length)?;
                return self.create_token($production, start);
            }};
        }

        match self.peek_characters(1)? {
            '(' => lex!(TokenKind::LeftParen, 1),
            ')' => lex!(TokenKind::RightParen, 1),
            '{' => lex!(TokenKind::LeftBrace, 1),
            '}' => lex!(TokenKind::RightBrace, 1),
            '[' => lex!(TokenKind::LeftBracket, 1),
            ']' => lex!(TokenKind::RightBracket, 1),
            ',' => lex!(TokenKind::Comma, 1),
            ';' => lex!(TokenKind::Semicolon, 1),

            ':' => {
                if let Some(':') = self.peek_characters(2) {
                    lex!(TokenKind::ColonColon, 2)
                }
                lex!(TokenKind::Colon, 1)
            }
            '.' => match self.peek_characters(2) {
                Some('*') => lex!(TokenKind::DotStar, 2),
                Some('&') => lex!(TokenKind::DotAmpersand, 2),
                Some('.')
                    if self
                        .peek_characters(3)
                        .is_some_and(|character| return character == '.') =>
                {
                    lex!(TokenKind::DotDotDot, 3)
                }
                _ => lex!(TokenKind::Dot, 1),
            },
            '=' => {
                if let Some('=') = self.peek_characters(2) {
                    lex!(TokenKind::EqualEqual, 2)
                }
                lex!(TokenKind::Equal, 1)
            }
            '+' => lex!(TokenKind::Plus, 1),
            '-' => lex!(TokenKind::Minus, 1),
            '*' => lex!(TokenKind::Star, 1),
            '/' => lex!(TokenKind::Slash, 1),
            '%' => lex!(TokenKind::Percent, 1),
            '&' => {
                if let Some('&') = self.peek_characters(2) {
                    lex!(TokenKind::AmpersandAmpersand, 2)
                }
                lex!(TokenKind::Ampersand, 1)
            }
            '|' => {
                if let Some('|') = self.peek_characters(2) {
                    lex!(TokenKind::PipePipe, 2)
                }
                lex!(TokenKind::Pipe, 1)
            }
            '!' => {
                if let Some('=') = self.peek_characters(2) {
                    lex!(TokenKind::ExclamationEqual, 2)
                }
                lex!(TokenKind::Exclamation, 1)
            }
            '^' => lex!(TokenKind::Caret, 1),
            '~' => lex!(TokenKind::Tilde, 1),
            '<' => match self.peek_characters(2) {
                Some('<') => lex!(TokenKind::LeftAngleLeftAngle, 2),
                Some('=') => lex!(TokenKind::LeftAngleEqual, 2),
                _ => lex!(TokenKind::LeftAngle, 1),
            },
            '>' => match self.peek_characters(2) {
                Some('>') => lex!(TokenKind::RightAngleRightAngle, 2),
                Some('=') => lex!(TokenKind::RightAngleEqual, 2),
                _ => lex!(TokenKind::RightAngle, 1),
            },

            'a'..='z' | 'A'..='Z' | '_' => {
                while let Some(character) = self.peek_characters(1) {
                    if character.is_ascii_alphanumeric() || character == '_' {
                        let _: char = self.next_characters(1)?;
                    } else {
                        break;
                    }
                }

                match self.source.get(start.range(&self.location))? {
                    "fn" => lex!(TokenKind::KeywordFn, 0),
                    "extern" => lex!(TokenKind::KeywordExtern, 0),
                    "let" => lex!(TokenKind::KeywordLet, 0),
                    "if" => lex!(TokenKind::KeywordIf, 0),
                    "else" => lex!(TokenKind::KeywordElse, 0),
                    "while" => lex!(TokenKind::KeywordWhile, 0),
                    "for" => lex!(TokenKind::KeywordFor, 0),
                    "return" => lex!(TokenKind::KeywordReturn, 0),
                    "break" => lex!(TokenKind::KeywordBreak, 0),
                    "continue" => lex!(TokenKind::KeywordContinue, 0),
                    "const" => lex!(TokenKind::KeywordConst, 0),
                    "true" => lex!(TokenKind::KeywordTrue, 0),
                    "false" => lex!(TokenKind::KeywordFalse, 0),
                    _ => lex!(TokenKind::Identifier, 0),
                }
            }

            '0'..='9' => {
                // Check for binary, octal, decimal, or hexadecimal integer literals
                match self.peek_characters(2) {
                    Some('b')
                        if self
                            .peek_characters(3)
                            .is_some_and(|character| return character.is_digit(2)) =>
                    {
                        let _: char = self.next_characters(2)?;

                        while let Some(character) = self.peek_characters(1) {
                            if character.is_digit(2) {
                                let _: char = self.next_characters(1)?;
                            } else {
                                break;
                            }
                        }

                        lex!(TokenKind::BinaryIntegerLiteral, 0);
                    }
                    Some('o')
                        if self
                            .peek_characters(3)
                            .is_some_and(|character| return character.is_digit(8)) =>
                    {
                        let _: char = self.next_characters(2)?;

                        while let Some(character) = self.peek_characters(1) {
                            if character.is_digit(8) {
                                let _: char = self.next_characters(1)?;
                            } else {
                                break;
                            }
                        }

                        lex!(TokenKind::OctalIntegerLiteral, 0);
                    }
                    Some('x')
                        if self
                            .peek_characters(3)
                            .is_some_and(|character| return character.is_ascii_hexdigit()) =>
                    {
                        let _: char = self.next_characters(2)?;

                        while let Some(character) = self.peek_characters(1) {
                            if character.is_ascii_hexdigit() {
                                let _: char = self.next_characters(1)?;
                            } else {
                                break;
                            }
                        }

                        lex!(TokenKind::HexadecimalIntegerLiteral, 0);
                    }
                    _ => {
                        while let Some(character) = self.peek_characters(1) {
                            if character.is_ascii_digit() {
                                let _: char = self.next_characters(1)?;
                            } else {
                                break;
                            }
                        }

                        match self.peek_characters(1) {
                            Some('.')
                                if self
                                    .peek_characters(2)
                                    .is_some_and(|character| character.is_ascii_digit()) =>
                            {
                                let _: char = self.next_characters(1)?;

                                while let Some(character) = self.peek_characters(1) {
                                    if character.is_ascii_digit() {
                                        let _: char = self.next_characters(1)?;
                                    } else {
                                        break;
                                    }
                                }

                                match self.peek_characters(1) {
                                    Some('e' | 'E') => {
                                        let _: char = self.next_characters(1)?;

                                        if let Some('+' | '-') = self.peek_characters(1) {
                                            let _: char = self.next_characters(1)?;
                                        }

                                        while let Some(character) = self.peek_characters(1) {
                                            if character.is_ascii_digit() {
                                                let _: char = self.next_characters(1)?;
                                            } else {
                                                break;
                                            }
                                        }

                                        lex!(TokenKind::FloatLiteral, 0);
                                    }
                                    _ => lex!(TokenKind::FloatLiteral, 0),
                                }
                            }
                            _ => lex!(TokenKind::DecimalIntegerLiteral, 0),
                        }
                    }
                }
            }

            '"' => {
                let _: char = self.next_characters(1)?;

                while let Some(character) = self.peek_characters(1) {
                    match character {
                        '\\' => {
                            let _: char = self.next_characters(1)?;
                            let _: char = self.next_characters(1)?;
                        }
                        '"' => {
                            let _: char = self.next_characters(1)?;
                            break;
                        }
                        _ => {
                            let _: char = self.next_characters(1)?;
                        }
                    }
                }

                lex!(TokenKind::StringLiteral, 0);
            }

            '\'' => {
                let _: char = self.next_characters(1)?;

                match self.peek_characters(1) {
                    Some('\\') => {
                        let _: char = self.next_characters(1)?;
                        let _: char = self.next_characters(1)?;
                    }
                    _ => {
                        let _: char = self.next_characters(1)?;
                    }
                }

                match self.peek_characters(1) {
                    Some('\'') => {
                        let _: char = self.next_characters(1)?;
                    }
                    _ => {
                        return None;
                    }
                }

                lex!(TokenKind::CharacterLiteral, 0);
            }

            _ => lex!(TokenKind::Unknown, 1),
        }
    }

    /// Build a token with the current state based on a token type and the
    /// starting position of the token
    fn create_token(
        &self,
        token_kind: TokenKind,
        start: Location<'file_name>,
    ) -> Option<Spanned<'file_name, Token<'source>>> {
        let span = Span::new(start, self.location);
        return Some(Spanned::new(
            span,
            Token::new(self.source.get(span.range())?, token_kind),
        ));
    }

    /// Skip any whitespace characters in the source code
    fn skip_whitespace(&mut self) {
        while let Some(character) = self.peek_characters(1) {
            if character.is_whitespace() {
                let _: char = self
                    .next_characters(1)
                    .expect("Failed to unwrap character already shown to exist");
            } else {
                break;
            }
        }
    }

    /// Peek into the source code by `count` characters
    ///
    /// Return the the `count` character from the current position if it exists,
    /// otherwise it returns `None`
    #[must_use]
    #[inline]
    fn peek_characters(&self, count: usize) -> Option<char> {
        let next = self.source.get(self.location.position()..)?;

        if next.len() < count {
            return None;
        }

        return next.chars().take(count).last();
    }

    /// Advance the lexer by `count` characters
    ///
    /// If `count` exceeds the remaining length then `None` is returned and
    /// the `Lexer` has not been updated.
    ///
    /// This will update the position based on UTF-8 character lengths, it will
    /// also update the line and column as appropriate
    #[must_use]
    #[inline]
    fn next_characters(&mut self, count: usize) -> Option<char> {
        let next = self.source.get(self.location.position()..)?;

        if next.len() < count {
            return None;
        }

        return next
            .chars()
            .take(count)
            .inspect(|character| {
                self.location.advance(*character);
            })
            .last();
    }
}

impl<'file_name, 'source> Iterator for Lexer<'file_name, 'source> {
    type Item = Spanned<'file_name, Token<'source>>;

    #[inline]
    fn next(&mut self) -> Option<Self::Item> {
        return self.next_token();
    }
}
