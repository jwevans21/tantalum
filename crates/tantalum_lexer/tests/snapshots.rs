use insta::assert_ron_snapshot;

use tantalum_lexer::Lexer;

macro_rules! snapshot_from_file {
    ($file_name:ident) => {
        #[test]
        fn $file_name() {
            let source = include_str!(concat!("../../examples/", stringify!($file_name), ".ta"));
            let lexer = Lexer::new(concat!(stringify!($file_name), ".ta"), source);

            let tokens = lexer.collect::<Vec<_>>();

            assert_ron_snapshot!(tokens);
        }
    };
    ($($file_name:ident),*) => {
        $(snapshot_from_file! { $file_name })*
    };
}

snapshot_from_file! {
    start,
    hello_world,
    conditionals,
    loops
}
