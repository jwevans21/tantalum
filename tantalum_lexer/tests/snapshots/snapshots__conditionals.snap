---
source: tantalum_lexer/tests/snapshots.rs
expression: tokens
snapshot_kind: text
---
[
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 0,
        line: 1,
        column: 1,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 2,
        line: 1,
        column: 3,
      ),
    ),
    lexeme: "fn",
    kind: KeywordFn,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 3,
        line: 1,
        column: 4,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 7,
        line: 1,
        column: 8,
      ),
    ),
    lexeme: "main",
    kind: Identifier,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 7,
        line: 1,
        column: 8,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 8,
        line: 1,
        column: 9,
      ),
    ),
    lexeme: "(",
    kind: LeftParen,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 8,
        line: 1,
        column: 9,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 9,
        line: 1,
        column: 10,
      ),
    ),
    lexeme: ")",
    kind: RightParen,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 9,
        line: 1,
        column: 10,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 10,
        line: 1,
        column: 11,
      ),
    ),
    lexeme: ":",
    kind: Colon,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 11,
        line: 1,
        column: 12,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 14,
        line: 1,
        column: 15,
      ),
    ),
    lexeme: "i32",
    kind: Identifier,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 15,
        line: 1,
        column: 16,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 16,
        line: 1,
        column: 17,
      ),
    ),
    lexeme: "{",
    kind: LeftBrace,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 21,
        line: 2,
        column: 5,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 24,
        line: 2,
        column: 8,
      ),
    ),
    lexeme: "let",
    kind: KeywordLet,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 25,
        line: 2,
        column: 9,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 26,
        line: 2,
        column: 10,
      ),
    ),
    lexeme: "a",
    kind: Identifier,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 26,
        line: 2,
        column: 10,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 27,
        line: 2,
        column: 11,
      ),
    ),
    lexeme: ":",
    kind: Colon,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 28,
        line: 2,
        column: 12,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 31,
        line: 2,
        column: 15,
      ),
    ),
    lexeme: "i32",
    kind: Identifier,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 32,
        line: 2,
        column: 16,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 33,
        line: 2,
        column: 17,
      ),
    ),
    lexeme: "=",
    kind: Equal,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 34,
        line: 2,
        column: 18,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 36,
        line: 2,
        column: 20,
      ),
    ),
    lexeme: "10",
    kind: DecimalIntegerLiteral,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 36,
        line: 2,
        column: 20,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 37,
        line: 2,
        column: 21,
      ),
    ),
    lexeme: ";",
    kind: Semicolon,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 42,
        line: 3,
        column: 5,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 45,
        line: 3,
        column: 8,
      ),
    ),
    lexeme: "let",
    kind: KeywordLet,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 46,
        line: 3,
        column: 9,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 47,
        line: 3,
        column: 10,
      ),
    ),
    lexeme: "b",
    kind: Identifier,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 47,
        line: 3,
        column: 10,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 48,
        line: 3,
        column: 11,
      ),
    ),
    lexeme: ":",
    kind: Colon,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 49,
        line: 3,
        column: 12,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 52,
        line: 3,
        column: 15,
      ),
    ),
    lexeme: "i32",
    kind: Identifier,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 53,
        line: 3,
        column: 16,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 54,
        line: 3,
        column: 17,
      ),
    ),
    lexeme: "=",
    kind: Equal,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 55,
        line: 3,
        column: 18,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 57,
        line: 3,
        column: 20,
      ),
    ),
    lexeme: "20",
    kind: DecimalIntegerLiteral,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 57,
        line: 3,
        column: 20,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 58,
        line: 3,
        column: 21,
      ),
    ),
    lexeme: ";",
    kind: Semicolon,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 63,
        line: 4,
        column: 5,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 66,
        line: 4,
        column: 8,
      ),
    ),
    lexeme: "let",
    kind: KeywordLet,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 67,
        line: 4,
        column: 9,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 68,
        line: 4,
        column: 10,
      ),
    ),
    lexeme: "c",
    kind: Identifier,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 68,
        line: 4,
        column: 10,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 69,
        line: 4,
        column: 11,
      ),
    ),
    lexeme: ":",
    kind: Colon,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 70,
        line: 4,
        column: 12,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 73,
        line: 4,
        column: 15,
      ),
    ),
    lexeme: "i32",
    kind: Identifier,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 74,
        line: 4,
        column: 16,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 75,
        line: 4,
        column: 17,
      ),
    ),
    lexeme: "=",
    kind: Equal,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 76,
        line: 4,
        column: 18,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 78,
        line: 4,
        column: 20,
      ),
    ),
    lexeme: "30",
    kind: DecimalIntegerLiteral,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 78,
        line: 4,
        column: 20,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 79,
        line: 4,
        column: 21,
      ),
    ),
    lexeme: ";",
    kind: Semicolon,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 85,
        line: 6,
        column: 5,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 87,
        line: 6,
        column: 7,
      ),
    ),
    lexeme: "if",
    kind: KeywordIf,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 88,
        line: 6,
        column: 8,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 89,
        line: 6,
        column: 9,
      ),
    ),
    lexeme: "a",
    kind: Identifier,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 90,
        line: 6,
        column: 10,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 91,
        line: 6,
        column: 11,
      ),
    ),
    lexeme: ">",
    kind: RightAngle,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 92,
        line: 6,
        column: 12,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 93,
        line: 6,
        column: 13,
      ),
    ),
    lexeme: "b",
    kind: Identifier,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 94,
        line: 6,
        column: 14,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 95,
        line: 6,
        column: 15,
      ),
    ),
    lexeme: "{",
    kind: LeftBrace,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 104,
        line: 7,
        column: 9,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 110,
        line: 7,
        column: 15,
      ),
    ),
    lexeme: "return",
    kind: KeywordReturn,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 111,
        line: 7,
        column: 16,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 112,
        line: 7,
        column: 17,
      ),
    ),
    lexeme: "a",
    kind: Identifier,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 112,
        line: 7,
        column: 17,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 113,
        line: 7,
        column: 18,
      ),
    ),
    lexeme: ";",
    kind: Semicolon,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 118,
        line: 8,
        column: 5,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 119,
        line: 8,
        column: 6,
      ),
    ),
    lexeme: "}",
    kind: RightBrace,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 120,
        line: 8,
        column: 7,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 124,
        line: 8,
        column: 11,
      ),
    ),
    lexeme: "else",
    kind: KeywordElse,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 125,
        line: 8,
        column: 12,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 127,
        line: 8,
        column: 14,
      ),
    ),
    lexeme: "if",
    kind: KeywordIf,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 128,
        line: 8,
        column: 15,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 129,
        line: 8,
        column: 16,
      ),
    ),
    lexeme: "b",
    kind: Identifier,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 130,
        line: 8,
        column: 17,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 131,
        line: 8,
        column: 18,
      ),
    ),
    lexeme: ">",
    kind: RightAngle,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 132,
        line: 8,
        column: 19,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 133,
        line: 8,
        column: 20,
      ),
    ),
    lexeme: "c",
    kind: Identifier,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 134,
        line: 8,
        column: 21,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 135,
        line: 8,
        column: 22,
      ),
    ),
    lexeme: "{",
    kind: LeftBrace,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 144,
        line: 9,
        column: 9,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 150,
        line: 9,
        column: 15,
      ),
    ),
    lexeme: "return",
    kind: KeywordReturn,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 151,
        line: 9,
        column: 16,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 152,
        line: 9,
        column: 17,
      ),
    ),
    lexeme: "b",
    kind: Identifier,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 152,
        line: 9,
        column: 17,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 153,
        line: 9,
        column: 18,
      ),
    ),
    lexeme: ";",
    kind: Semicolon,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 158,
        line: 10,
        column: 5,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 159,
        line: 10,
        column: 6,
      ),
    ),
    lexeme: "}",
    kind: RightBrace,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 160,
        line: 10,
        column: 7,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 164,
        line: 10,
        column: 11,
      ),
    ),
    lexeme: "else",
    kind: KeywordElse,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 165,
        line: 10,
        column: 12,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 166,
        line: 10,
        column: 13,
      ),
    ),
    lexeme: "{",
    kind: LeftBrace,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 175,
        line: 11,
        column: 9,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 181,
        line: 11,
        column: 15,
      ),
    ),
    lexeme: "return",
    kind: KeywordReturn,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 182,
        line: 11,
        column: 16,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 183,
        line: 11,
        column: 17,
      ),
    ),
    lexeme: "c",
    kind: Identifier,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 183,
        line: 11,
        column: 17,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 184,
        line: 11,
        column: 18,
      ),
    ),
    lexeme: ";",
    kind: Semicolon,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 189,
        line: 12,
        column: 5,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 190,
        line: 12,
        column: 6,
      ),
    ),
    lexeme: "}",
    kind: RightBrace,
  ),
  Token(
    span: Span(
      start: Location(
        file_name: "conditionals.ta",
        position: 191,
        line: 13,
        column: 1,
      ),
      end: Location(
        file_name: "conditionals.ta",
        position: 192,
        line: 13,
        column: 2,
      ),
    ),
    lexeme: "}",
    kind: RightBrace,
  ),
]
